### PR TITLE
ramips: add support for HiLink HLK-7621A evaluation board

### PR DESCRIPTION
**Pull request by paraka**
Specifications:
   - SoC: MediaTek MT7621AT
   - RAM: 256 MB (DDR3)
   - Flash: 32 MB SPI NOR 44MHz
   - Switch: 1 WAN, 4 LAN (Gigabit)
   - LEDs: 1 WAN, 4 LAN (controlled by PHY)
   - USB Ports: 1 x USB2, 1 x USB3
   - WLAN: 1 x 2.4, 5 GHz 866Mbps (MT7612E)
   - Button: 1 button (reset)
   - UART Serial: UART1 as console : 57600 baud
   - Power: 12V, 1A

Installation:

- Update openWRT firmware by internal GNUBEE uboot: https://github.com/gnubee-git/GnuBee-MT7621-uboot
- By HTTP: Initial uboot address is http://10.10.10.123, your address needs to be 10.10.10.x, and mask 255.255.255.0.
- By TFTP: Uboot is in client mode, the address of the firmware must be tftp://10.10.10.3/uboot.bin

Recovery:
- Manufacturer provides MTK OpenWrt 14.07 source code, compile then flash it by uboot.

HLK-7621A is a stamp hole package module for embedded development, users have to design IO boards to use it.

 Mac addresses:
- Ethernet mac address is only located in uboot-env partition and can be retrieved as follows:
       # mtd_get_mac_ascii u-boot-env ethaddr
       03:17:73:ab:cd:ef
      Since this address is obviously a placeholder should not be used.
      We will set factory partition 0x8004 offset as mac start value to have a fixed address
  instead of random one since we cannot do anything better if vendor is not setting up
  properly this. Hence ethernet mac will be: 8c:88:2b:00:00:1b.
- Wifi mac address is located at '/sys/class/ieee80211/phy0/macaddress':
       # cat /sys/class/ieee80211/phy0/macaddress
       f8:62:aa:50:11:a8
      This is already properly assigned so nothing is necessary to be done.

User Manual (Chinese & English):
[HLK-7621A用户手册V1.5.pdf](https://github.com/openwrt/openwrt/files/5884699/HLK-7621A.V1.5.pdf)

Signed-off-by: Chen Yijun <cyjason[at]bupt[dot]edu[dot]cn>
[add keys and pcie nodes to properly support evaluation board]
Signed-off-by: Sergio Paracuellos <sergio.paracuellos[at]gmail[dot]com>

NOTE: This superseds the following PR: https://github.com/openwrt/openwrt/pull/3817

### CODE REVIEW

**Comment by adschm** [comment]
Duplicate of #3817.

**Comment by paraka** [comment]
@adschm  Ok, so I get this closed because of duplication but there is no feedback/response at all in #3817. I really want this to be added to openwrt code, so.. how should I proceed? Thanks in advance for clarification.

**Comment by adschm** [comment]
I see your point. It's just a pain to have multiple PRs for the same device, as typically this also leads to duplicate work and makes looking up history harder.

**Comment by paraka** [comment]
> I see your point. It's just a pain to have multiple PRs for the same device, as typically this also leads to duplicate work and makes looking up history harder.

Thanks for clarification, Adrian. As you can see I had added the Sign-on-by of the previous author also in the patch since it was my starting point for adding all the rest of the stuff.

**Comment by MichaelWarneke** [comment]
Is there anything missing for this PR to get merged?
I used it now on my device for quite some time and it looks very stable.
Can assist in any issues.

**Comment by paraka** [comment]
@MichaelWarneke: I also don't know. I think is totally ready to be merged. 
@adschm: When these kind of PRs are merged into the tree?

File: target/linux/ramips/image/mt7621.mk
Comment (blocktrron) [inline target/linux/ramips/image/mt7621.mk:569]: missing indent.

File: target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts
Comment (blocktrron) [inline target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts:106]: Add a trailing newline.

**Review by blocktrron — COMMENTED** [review]
Let's make some progress here.

File: target/linux/ramips/image/mt7621.mk
Comment (paraka) [reply target/linux/ramips/image/mt7621.mk]: True. Fixed now.

File: target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts
Comment (paraka) [reply target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts]: Done

**Comment by paraka** [comment]
@blocktrron Thanks for the review :)

**Review by MichaelWarneke — APPROVED** [review]

**Comment by paraka** [comment]
@MichaelWarneke Thanks!

File: target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts
Comment (adschm) [inline target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts:65]: This can be dropped, it's enabled by default.

File: target/linux/ramips/image/mt7621.mk
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:617]: Please make this consistent with the "model" in DTS.

**Comment by adschm** [comment]
> Manufacturer also provides an evaluation board for this module with the following interfaces:
> - 12V1A power
> - pcie peripheral interface
> - usb2
> - usb3
> - uart part
> - uart-usb
> - 5 GB ports (4 lan ports and 1 wan).
> - 1 default button.

This reads like it's some other, irrelevant board. But from the DTS, it's the board you built support for?
Why not add it to the specs so this is obvious ...

Between the two Signed-off-by, you should describe all the changes you made since the original version, e.g. https://github.com/openwrt/openwrt/commit/ae9c5cd37bf5c08452f314b54afa963a00bdde30

What about MAC addresses? Are the all random at the moment?

File: target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts
Comment (paraka) [reply target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts:65]: Are you sure? I had to add this explicitly to be able to use the usb ports in the board. It did not working without this line.

File: target/linux/ramips/image/mt7621.mk
Comment (paraka) [reply target/linux/ramips/image/mt7621.mk:617]: Will change into  "HLK-7621A evaluation board" as it is in dts file.

**Comment by paraka** [comment]
> > Manufacturer also provides an evaluation board for this module with the following interfaces:
> > 
> > * 12V1A power
> > * pcie peripheral interface
> > * usb2
> > * usb3
> > * uart part
> > * uart-usb
> > * 5 GB ports (4 lan ports and 1 wan).
> > * 1 default button.
> 
> This reads like it's some other, irrelevant board. But from the DTS, it's the board you built support for?
> Why not add it to the specs so this is obvious ...

Yes, it is the board I built support for. What do you mean with "add it to the specs"?

> 
> Between the two Signed-off-by, you should describe all the changes you made since the original version, e.g. [ae9c5cd](https://github.com/openwrt/openwrt/commit/ae9c5cd37bf5c08452f314b54afa963a00bdde30)

I see, thanks. Will add it.

> 
> What about MAC addresses? Are the all random at the moment?

Yes, should I say "MAC addresses are random" in the commit message?

File: target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts
Comment (adschm) [reply target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts:65]: Maybe you tested with an older stable release?

Comment (adschm) [reply target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts:65]: https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/dts/mt7621.dtsi#L371

**Comment by adschm** [comment]
> Yes, it is the board I built support for. What do you mean with "add it to the specs"?

Well, you have "Specifications" on top and then this list on the bottom. All of it should just be "Specifications" on the top if I understand correctly.

**Comment by adschm** [comment]
> Yes, should I say "MAC addresses are random" in the commit message?

Even better would be to not have random addresses at all, but assign them properly.

If you really don't have any predefined MAC addresses on the board, then please add a comment about this to the commit message (including the situation with WiFi).

**Comment by paraka** [comment]
> > Manufacturer also provides an evaluation board for this module with the following interfaces:
> > 
> > * 12V1A power
> > * pcie peripheral interface
> > * usb2
> > * usb3
> > * uart part
> > * uart-usb
> > * 5 GB ports (4 lan ports and 1 wan).
> > * 1 default button.
> 
> This reads like it's some other, irrelevant board. But from the DTS, it's the board you built support for?
> Why not add it to the specs so this is obvious ...

Yes, it is the board I built support for. What do you mean with "add it to the specs"?

> 
> Between the two Signed-off-by, you should describe all the changes you made since the original version, e.g. [ae9c5cd](https://github.com/openwrt/openwrt/commit/ae9c5cd37bf5c08452f314b54afa963a00bdde30)

I see, thanks. Will add it.

> 
> What about MAC addresses? Are the all random at the moment?

Yes, should I say "MAC addresses are random" in the commit message?



> > Yes, it is the board I built support for. What do you mean with "add it to the specs"?
> 
> Well, you have "Specifications" on top and then this list on the bottom. All of it should just be "Specifications" on the top if I understand correctly.

Ah ok, got it. Thanks for the explanation.

File: target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts
Comment (paraka) [reply target/linux/ramips/dts/mt7621_hilink_hlk-7621a-evb.dts:65]: You are right. I have just tested with last code in master branch and works. Thanks, will drop this.

**Comment by paraka** [comment]
> > Yes, should I say "MAC addresses are random" in the commit message?
> 
> Even better would be to not have random addresses at all, but assign them properly.
> 
> If you really don't have any predefined MAC addresses on the board, then please add a comment about this to the commit message (including the situation with WiFi).

So after looking into the FLASH the only information about the ethernet mac is in u-boot-env as an ascii string. There is nothing about the wifi's MAC. Can I derive also the wifi mac as follows? Is this a valid approach?

diff --git a/target/linux/ramips/mt7621/base-files/etc/board.d/02_network b/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
index aa1d644938..044de868bf 100644
--- a/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
+++ b/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
@@ -121,6 +121,10 @@ ramips_setup_macs()
                wan_mac=$(mtd_get_mac_ascii u-boot-env wanaddr)
                label_mac=$wan_mac
                ;;
+       hilink,hlk-7621a-evb)
+               lan_mac=$(mtd_get_mac_ascii u-boot-env ethaddr)
+               wan_mac=$(macaddr_add "$lan_mac" 1)
+               ;;
        hiwifi,hc5962)
                lan_mac=$(mtd_get_mac_ascii bdinfo "Vfac_mac ")
                label_mac=$lan_mac

File: target/linux/ramips/image/mt7621.mk
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:619]: No wifi device packages?

**Comment by adschm** [comment]
> And I can retrieve it using:
> 
> ```
> root@OpenWrt:/# . /lib/functions/system.sh 
> root@OpenWrt:/# mtd_get_mac_ascii u-boot-env ethaddr
> 03:17:73:ab:cd:ef
> ```

Technically, you would use the lower part of https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/board.d/02_network

However, what you found here is obviously a placeholder address and should not be used.

> 
> Also the wifi's mac address is only located in:
> 
> ```
> cat /sys/class/ieee80211/phy0/macaddress 
> f8:62:aa:50:11:a8
> ```

Okay, this looks like a real address. But since it's already properly assigned you don't have to do anything about it.

**Comment by paraka** [comment]
> > And I can retrieve it using:
> > ```
> > root@OpenWrt:/# . /lib/functions/system.sh 
> > root@OpenWrt:/# mtd_get_mac_ascii u-boot-env ethaddr
> > 03:17:73:ab:cd:ef
> > ```
> 
> Technically, you would use the lower part of https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/mt7621/base-files/etc/board.d/02_network
> 
> However, what you found here is obviously a placeholder address and should not be used.
> 
> > Also the wifi's mac address is only located in:
> > ```
> > cat /sys/class/ieee80211/phy0/macaddress 
> > f8:62:aa:50:11:a8
> > ```
> 
> Okay, this looks like a real address. But since it's already properly assigned you don't have to do anything about it.

Ok, thanks. I will add a comment about this two mac things in the commit message and don't do anything, then.

File: target/linux/ramips/image/mt7621.mk
Comment (paraka) [reply target/linux/ramips/image/mt7621.mk:619]: Ok, so I can add here also the device packages for the wifi, yes. I have assigned them manually in menu config :). Thanks for pointing this out. Will add:

DEVICE_PACKAGES += kmod-mt76x2 kmod-usb3

**Comment by paraka** [comment]
@adschm Please, let me know if there is something missing to be changed here. Thank you very much for all of review comments.

**Comment by adschm** [comment]
> Update firmware by internal GNUBEE uboot: https://github.com/gnubee-git/GnuBee-MT7621-uboot

Err, this needs to replace the bootloader? That's typically a no-go in this target ...

And since we seem to have a MAC address for Wifi, please check whether that's in 0x8004. If yes, please also assign it to ethernet (and add a comment to the commit message that we deliberately chose this instead of a random address, though the vendor did not).

**Comment by paraka** [comment]
> > Update firmware by internal GNUBEE uboot: https://github.com/gnubee-git/GnuBee-MT7621-uboot
> 
> Err, this needs to replace the bootloader? That's typically a no-go in this target ...

No, this means bootloader is already installed in the board and using it you can update openwrt firmware. You can also ssh/serial access to the board and use sysupgrade command from there to upgrade the openwrt firmware also.

> 
> And since we seem to have a MAC address for Wifi, please check whether that's in 0x8004. If yes, please also assign it to ethernet (and add a comment to the commit message that we deliberately chose this instead of a random address, though the vendor did not).

This is what I get from factory partition 0x8004:

00008000  62 76 01 00 8c 88 2b 00  00 1b 12 76 c3 14 00 00  |bv....+....v....|
00008010  00 00 12 76 c3 14 00 00  00 00 ff ff ff ff ff ff  |...v............|
00008020  ff ff 37 d8 9d 40 ff 7f  fd 9b ff ff ff ff ff ff  |..7..@..........|
00008030  ff ff ff ff 22 fd 0a 00  ff ff b1 01 00 00 00 00  |...."...........|
00008040  00 00 ff d3 00 8a 00 00  00 8a 00 00 00 8b e0 01  |................|
00008050  00 86 88 57 47 b8 00 00  1a 00 00 00 00 00 1a 00  |...WG...........|
00008060  00 00 7a 06 1a 00 00 7c  0a 19 80 c1 ff 04 1c 80  |..z....|........|
00008070  c4 ff 12 20 80 c0 ff f8  1c 80 c5 ff 14 21 80 c1  |... .........!..|
00008080  ff 0a 1a 00 00 80 09 1f  80 c2 ff 20 21 80 c3 ff  |........... !...|
00008090  20 24 80 c0 ff 0a 20 80  c5 ff 0a 24 80 c1 ff 32  | $.... ....$...2|
000080a0  c3 c3 c8 c8 c8 c8 cc ca  c4 00 cc ca c6 00 00 00  |................|
000080b0  00 00 cc cc c8 00 00 00  00 00 00 00 00 00 81 81  |................|
000080c0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|

Do you want me to use "8c::88::2b::00::00::1b" as MAC address? I don't really understand if this is what you want me to do. Thanks for clarification.

**Comment by MichaelWarneke** [comment]
What needs to be done to get this merged?

**Comment by paraka** [comment]
> What needs to be done to get this merged?

Not sure what might be missing yet. I have addressed all the comments, so it should be enough to get this merged.

**Comment by paraka** [comment]
Gentle ping for this request.

**Comment by MichaelWarneke** [comment]
Any updates on this?

**Comment by paraka** [comment]
> Any updates on this?

Honestly, I don't know. I don't understand why six months are needed to merge a pull request. I normally contribute with patches to the mainline linux kernel and have never seen something similar. I guess everybody is busy with other stuff. I hope one day this will be merged. Thanks @MichaelWarneke for the interest so far.

**Comment by paraka** [comment]
Converted from mtd-mac-address to nvmem implementation

**Comment by adschm** [comment]
I don't think the address at 0x8004 is real. The 00:00 for the right are very unlikely.

Does anybody have information about the OEM/default firmware? Will ethernet get a random address there?

Without further info, I'd use the following for the commit message:

```
    MAC addresses:
    - u-boot-env contains a placeholder address:
      > mtd_get_mac_ascii u-boot-env ethaddr
      03:17:73:ab:cd:ef
    - phy0 gets a valid-looking address:
       # cat /sys/class/ieee80211/phy0/macaddress
      f8:62:aa:**:**:a8
    - Calibration data for &pcie2 contains a valid address, however the
      zeros in the right half look like it's not real:
      8c:88:2b:00:00:1b
    - Since it's an evaluation board and there is no solid information
      about the MAC address assignment, the ethernet MAC address is left random.
```

Apart from that, note that with the keys added we also need to add a proper state_default node. I'd also like a full dump of the factory partition so I can check for the other MAC address.

Does anybody know where the phy0 MAC address is set from?

**Comment by paraka** [comment]
> I don't think the address at 0x8004 is real. The 00:00 for the right half are very unlikely.

You ask me to use what is in 0x8004 since vendor does not provide it, just to avoid a random value for ethernets.
> 
> Does anybody have information about the OEM/default firmware? Will ethernet get a random address there?

No, I don't have that information.

> 
> Without further info, I'd use the following for the commit message:
> 
> ```
>     MAC addresses:
>     - u-boot-env contains a placeholder address:
>       > mtd_get_mac_ascii u-boot-env ethaddr
>       03:17:73:ab:cd:ef
>     - phy0 gets a valid-looking address:
>        # cat /sys/class/ieee80211/phy0/macaddress
>       f8:62:aa:**:**:a8
>     - Calibration data for &pcie2 contains a valid address, however the
>       zeros in the right half look like it's not real:
>       8c:88:2b:00:00:1b
>     - Since it's an evaluation board and there is no solid information
>       about the MAC address assignment, the ethernet MAC address is left random.
> ```

Okay I will set this in the commit message.

> 
> Apart from that, note that with the keys added we also need to add a proper state_default node. I'd also like a full dump of the factory partition so I can check for the other MAC address.

Here you are:

root@OpenWrt:/# cat /proc/mtd 
dev:    size   erasesize  name
mtd0: 00030000 00010000 "u-boot"
mtd1: 00010000 00010000 "u-boot-env"
mtd2: 00010000 00010000 "factory"
mtd3: 01fb0000 00010000 "firmware"
mtd4: 00294f14 00010000 "kernel"
mtd5: 01d1b0ec 00010000 "rootfs"
mtd6: 01960000 00010000 "rootfs_data"

root@OpenWrt:/# hexdump -C /dev/mtd2
00000000  28 76 00 02 8c 88 2b 00  00 18 00 00 00 00 00 00  |(v....+.........|
00000010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
00000020  00 00 00 00 20 00 00 00  8c 88 2b 00 00 19 8c 88  |.... .....+.....|
00000030  2b 00 00 1a 22 34 00 20  ff ff 00 01 00 00 00 00  |+..."4. ........|
00000040  00 00 22 00 00 00 00 00  30 00 00 00 00 00 00 00  |..".....0.......|
00000050  81 00 00 94 40 b6 c0 ca  00 00 00 00 40 ca 00 00  |....@.......@...|
00000060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
000000a0  ca ca c7 ca c6 c0 c0 cd  d0 cc c7 c4 c0 81 00 00  |................|
000000b0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
000000c0  00 00 00 00 00 00 00 00  00 1a 22 2a 31 35 01 35  |.........."*15.5|
000000d0  39 40 46 4d 7f 7f 7f 00  00 00 00 00 00 00 00 00  |9@FM............|
000000e0  11 1d 11 1d 1c 35 1c 35  1e 35 1e 35 17 19 17 19  |.....5.5.5.5....|
000000f0  02 00 00 00 b0 c8 80 88  0a 00 00 00 00 00 00 00  |................|
00000100  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00000120  00 00 00 00 00 00 00 00  00 00 00 00 00 00 77 00  |..............w.|
00000130  11 1d 11 1d 15 7f 15 7f  17 7f 17 7f 10 3b 10 3b  |.............;.;|
00000140  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00008000  62 76 01 00 8c 88 2b 00  00 1b 12 76 c3 14 00 00  |bv....+....v....|
00008010  00 00 12 76 c3 14 00 00  00 00 ff ff ff ff ff ff  |...v............|
00008020  ff ff 37 d8 9d 40 ff 7f  fd 9b ff ff ff ff ff ff  |..7..@..........|
00008030  ff ff ff ff 22 fd 0a 00  ff ff b1 01 00 00 00 00  |...."...........|
00008040  00 00 ff d3 00 8a 00 00  00 8a 00 00 00 8b e0 01  |................|
00008050  00 86 88 57 47 b8 00 00  1a 00 00 00 00 00 1a 00  |...WG...........|
00008060  00 00 7a 06 1a 00 00 7c  0a 19 80 c1 ff 04 1c 80  |..z....|........|
00008070  c4 ff 12 20 80 c0 ff f8  1c 80 c5 ff 14 21 80 c1  |... .........!..|
00008080  ff 0a 1a 00 00 80 09 1f  80 c2 ff 20 21 80 c3 ff  |........... !...|
00008090  20 24 80 c0 ff 0a 20 80  c5 ff 0a 24 80 c1 ff 32  | $.... ....$...2|
000080a0  c3 c3 c8 c8 c8 c8 cc ca  c4 00 cc ca c6 00 00 00  |................|
000080b0  00 00 cc cc c8 00 00 00  00 00 00 00 00 00 81 81  |................|
000080c0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000080f0  ff ff 0f 09 0d 10 53 00  24 00 00 00 00 00 00 00  |......S.$.......|
00008100  0f ca 74 c5 e8 07 30 3d  01 b0 08 26 00 0e 04 15  |..t...0=...&....|
00008110  00 8a 00 40 00 00 00 08  00 9d 08 00 12 c0 00 00  |...@............|
00008120  08 20 04 2a 90 00 00 24  01 04 54 08 d0 a0 28 20  |. .*...$..T...( |
00008130  ff ff ff ff ff ff ff 08  ff ff ff ff ff ff 00 00  |................|
00008140  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
000081e0  c0 81 82 c3 04 45 46 07  08 09 ff ff ff ff ff ff  |.....EF.........|
000081f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
0000ff80  ff ff ff ff 32 30 31 36  30 38 32 35 71 68 20 20  |....20160825qh  |
0000ff90  20 20 20 20 20 20 20 20  ff ff ff ff ff ff ff ff  |        ........|
0000ffa0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
0000ffe0  ff ff ff ff ff ff ff ff  c4 36 55 75 00 4f c4 36  |.........6Uu.O.6|
0000fff0  55 75 00 50 ff ff ff ff  ff ff ff ff ff ff ff ff  |Uu.P............|
00010000
> 
> Does anybody know where the phy0 MAC address is set from? Would we maybe need a proper &wmac node for that?

No idea, but since it looks like is a valid mac address you said me before that there is nothing extra needed to do.

Thanks.

**Comment by paraka** [comment]
'state_default' added to device tree. Re-do the commit message in the way I have been asked to do.

**Comment by adschm** [comment]
In principle, the MAC address assignment can only be done properly with information from OEM and/or multiple devices.

Since this is an evaluation board (and it's unrealistic to what for such additional information), I will nevertheless merge it in the current incomplete state, so we get it off the table. Thanks for dealing with some of the loose ends of the original PR.

**Comment by paraka** [comment]
> In principle, the MAC address assignment can only be done properly with information from OEM and/or multiple devices.
> 
> Since this is an evaluation board (and it's unrealistic to wait for such additional information), I will nevertheless merge it in the current incomplete state, so we get it off the table. Thanks for dealing with some of the loose ends of the original PR.

So it is preferred to don't have support at all for people that use this board for evaluation. I don't understand and not agree at all with your decision but you have write access to the repository so I guess these are your rules. Thanks for letting me know. For the future, to avoid to waste other people's time (as I have been wasting it adapting this stuff to all the nvmem and other changes) don't let more than six months to pass if you have clear enough that you are going to close the PR at the end.

**Comment by adschm** [comment]
> > In principle, the MAC address assignment can only be done properly with information from OEM and/or multiple devices.
> > Since this is an evaluation board (and it's unrealistic to wait for such additional information), I will nevertheless merge it in the current incomplete state, so we get it off the table. Thanks for dealing with some of the loose ends of the original PR.
> 
> So it is preferred to don't have support at all for people that use this board for evaluation. I don't understand and not agree at all with your decision but you have write access to the repository so I guess these are your rules. Thanks for letting me know. For the future, to avoid to waste other people's time (as I have been wasting it adapting this stuff to all the nvmem and other changes) don't let more than six months to pass if you have clear enough that you are going to close the PR at the end.

I think you misunderstood me. I said I AM GOING TO MERGE it, it's already in my local tree.
I'm just closing the PR here now that the subject is resolved; I will push the commit in the next few days.

**Comment by champtar** [comment]
@paraka It can be confusing but for this repo devs do not use Github merge feature, they pull / review / tests on their computer, then push to https://git.openwrt.org/?p=openwrt/openwrt.git;a=summary

**Comment by paraka** [comment]
> > So it is preferred to don't have support at all for people that use this board for evaluation. I don't understand and not agree at all with your decision but you have write access to the repository so I guess these are your rules. Thanks for letting me know. For the future, to avoid to waste other people's time (as I have been wasting it adapting this stuff to all the nvmem and other changes) don't let more than six months to pass if you have clear enough that you are going to close the PR at the end.
> 
> I think you misunderstood me. I said I AM GOING TO MERGE it, it's already in my local tree. I'm just closing the PR here now that the subject is resolved; I will push the commit in the next few days.

Sorry, yes I misunderstood you. Thanks for letting me know. I understand that if you don't have OEM configuration the rules might be to don't merge the PR. What I did not understood was that I was not tell about that in first place. That was my misunderstood and because of that I just share with you my feelings. I apologize since it seems  for that.

> 
> Count yourself lucky that this is an evaluation board, because otherwise I would indeed have declined merging it without proper MAC address setup (which can only come from OEM configuration).

Understood.

> 
> Note that you are quite quick with ranting about being treated unfairly. This does _not_ help you at all. Just keep in mind that reviewers will have to choose between a lot of open PRs to take care of; why would they choose yours if you tell them how stupid they are ...

I was not telling stupid to anyone. I only though that the PR was closed after more than six months because of MAC assignation which was a thing that was on the table from the beginning. Because of this I though I was been wasting my time (nothing to do with being treated unfairly at all, rules are rules but must be from the beginning) and get a bit angry. I respect a lot the work of the reviewers, not only in openWRT but in any single project since I also review patches in linux kernel. So, again, before, I misunderstood you and I apologize for that.

**Comment by paraka** [comment]
> @paraka It can be confusing but for this repo devs do not use Github merge feature, they pull / review / tests on their computer, then push to https://git.openwrt.org/?p=openwrt/openwrt.git;a=summary

I didn't know. I am understanding now. Thanks.

**Comment by NikZhukov** [comment]
Today's Snapshot is r18503.
After the firmware during the download, multiple errors: jffs2: jffs2_scan_erase block(): Magic bitmask 0x1985 not found at ... instead and jffs2: Newly-erased block contained word ....... Not sure if this is related to the snapshot, since it was collected for the first time. But these are the same errors on the downloaded snapshot. Tell me what's wrong.

**Comment by cyijun** [comment]
> Today's Snapshot is r18503. After the firmware during the download, multiple errors: _jffs2: jffs2_scan_erase block(): Magic bitmask 0x1985 not found at ... instead_ and _jffs2: Newly-erased block contained word ......._ Not sure if this is related to the snapshot, since it was collected for the first time. But these are the same errors on the downloaded snapshot. Tell me what's wrong.

@NikZhukov Hi I am the other author of this pull request, I have not met these problems yet. Would you like to try flashing another bootloader? I am using a closed source but free to use bootloader, you can find it [here](https://github.com/cyijun/MT7621A-Router-Design) .

**Comment by rma-x** [comment]
I recently got two of these boards (but have only one of them at hand right now), and it looks like I can shed some light at the MAC address discussion from above.

First of all, this board does not have builtin WiFi hardware, so I think it shouldn't reference any wifi in the dts file. WiFi has to be added via the PCIe slot, but other PCIe cards work there as well, e.g. I put a GbE card in to have an additoonal Ethernet port, which works fine after installing the respective kmod package and without any need to tinker with the device tree, so I guess the same holds true for WiFi cards in the PCIe slot.

Secondly, I compared the factory partition dump from above with my device and found the only difference to be what looks like two subsequent MAC addresses (40:d6:3c:c1:00:48, and 40:d6:3c:c1:00:49) at partition offset 0xe000, probably meant to be used for the LAN and WAN interfaces:

```
 000081e0  c0 81 82 c3 04 45 46 07  08 09 ff ff ff ff ff ff  |.....EF.........|
 000081f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
 *
+0000e000  40 d6 3c c1 00 48 40 d6  3c c1 00 49 ff ff ff ff  |@.<..H@.<..I....|
+0000e010  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
+*
 0000ff80  ff ff ff ff 32 30 31 36  30 38 32 35 71 68 20 20  |....20160825qh  |
 0000ff90  20 20 20 20 20 20 20 20  ff ff ff ff ff ff ff ff  |        ........|
 0000ffa0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
```
Next time I get my hands on the second board, I'll compare that one as well to see if the MAC addresses are really individual.

**Pull request closed without merging**
